jps-plugin/src/org/intellij/erlang/jps/builder/ErlangPrepareDependenciesBuilder.java

```diff
@@ -0,0 +1,270 @@
+                                          @NotNull List<String> dirtyErlangFilePaths) {
+    SortedModuleDependencyGraph graph = new SortedModuleDependencyGraph(dependencyTree);
+    Set<String> allDirtyFiles = getAllDirtyFiles(dirtyErlangFilePaths, GraphGenerator.create(graph));
+    ErlangModuleBuildOrders buildOrders = createBuildOrdersForDirtyErlangModules(project, dependencyTree.myNodes, allDirtyFiles);
+    try {
+      Document serializedBuildOrders = new Document(XmlSerializer.serialize(buildOrders, new SkipDefaultValuesSerializationFilters()));
+      File dataStorageRoot = context.getProjectDescriptor().dataManager.getDataPaths().getDataStorageRoot();
+      File depsConfigFile = new File(dataStorageRoot, ErlangBuilder.DEPENDENCIES_CONFIG_FILE_PATH);
+      //noinspection ResultOfMethodCallIgnored
+      depsConfigFile.getParentFile().mkdirs();
+      JDOMUtil.writeDocument(serializedBuildOrders, depsConfigFile, SystemProperties.getLineSeparator());
+    }
+    catch (XmlSerializationException e) {
+      addPrepareDependenciesFailedMessage(context);
+    }
+    catch (IOException e) {
+      addPrepareDependenciesFailedMessage(context);
+    }
+  }
+
+  private static void addPrepareDependenciesFailedMessage(@NotNull CompileContext context) {
+    context.processMessage(new CompilerMessage(NAME, BuildMessage.Kind.WARNING, "Some error has occurred while prepare dependencies." +
+                                                                                " Project will be built without using the information about dependencies and no incremental."));
+  }
+
+  private static Set<String> getAllDirtyFiles(@NotNull List<String> dirtyModules,
+                                              @NotNull GraphGenerator<String> graph) {
+    Set<String> dirtyNodes = ContainerUtil.newHashSet();
+    for (String node : dirtyModules) {
+      if (node != null) {
+        findDirtyNodes(node, graph, dirtyNodes);
+      }
+    }
+    return dirtyNodes;
+  }
+
+  private static void findDirtyNodes(@NotNull String node,
+                                     @NotNull GraphGenerator<String> graph,
+                                     @NotNull Set<String> dirtyNodes) {
+    if (dirtyNodes.contains(node)) return;
+    dirtyNodes.add(node);
+    Iterator<String> childIterator = graph.getOut(node);
+    while (childIterator.hasNext()) {
+      findDirtyNodes(childIterator.next(), graph, dirtyNodes);
+    }
+  }
+
+  @NotNull
+  private static ErlangModuleBuildOrders createBuildOrdersForDirtyErlangModules(@NotNull JpsProject project,
+                                                                                @NotNull List<ErlangDependencyTreeNode> nodes,
+                                                                                @NotNull Set<String> dirtySortedErlangFiles) {
+    return new BuildOrdersCreator(project, nodes).createBuildOrderForDirtyFiles(dirtySortedErlangFiles);
```

Comment:
There we have it again. When you wrote initial version of incremental compilation, you had the exact same case in ErlangPrepareDependenciesCompileTask - it's a recurring pattern - you're trying to extract an object for something which essentially isn't an object - it's a function, a single static method. 

It can be convenient to have a wrapping object, though, in order not to pass around a lot of state needed for computation. In this case, just hide the object creation from outside world as we did it in ErlangPrepareDependenciesCompileTask.